beater/middleware.go

```diff
@@ -0,0 +1,31 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"github.com/elastic/apm-server/beater/request"
+)
+
+type middleware func(request.Handler) request.Handler
+
+func withMiddleware(h request.Handler, m ...middleware) request.Handler {
+	for i := len(m) - 1; i >= 0; i-- {
```

Comment:
I was surprised that no tests failed when flipping the order the other direction

Reply:
Yes we are lacking some integration tests that actually test the middleware. I am working on adding them in the next PR.